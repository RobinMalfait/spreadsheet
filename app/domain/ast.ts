import type { Location } from '~/domain/expression'

export enum AstKind {
  CELL = 'CELL',
  RANGE = 'RANGE',
  FUNCTION = 'FUNCTION',
  NUMBER_LITERAL = 'NUMBER_LITERAL',
  STRING_LITERAL = 'STRING_LITERAL',
  BINARY_EXPRESSION = 'BINARY_EXPRESSION',
}

interface Span {
  span: {
    start: number
    end: number
  }
}

export interface AstCell extends Span {
  kind: AstKind.CELL
  name: string
  raw: string
  loc: Location
}

export interface AstCellRange extends Span {
  kind: AstKind.RANGE
  raw: string
  start: AstCell
  end: AstCell
}

export interface AstFunction extends Span {
  kind: AstKind.FUNCTION
  name: string
  args: AST[]
}

interface AstNumberLiteral extends Span {
  kind: AstKind.NUMBER_LITERAL
  value: number
}

interface AstStringLiteral extends Span {
  kind: AstKind.STRING_LITERAL
  value: string
}

interface AstBinaryExpression extends Span {
  kind: AstKind.BINARY_EXPRESSION
  operator: BinaryExpressionOperator
  lhs: AST
  rhs: AST
}

export type AST =
  | AstCell
  | AstCellRange
  | AstFunction
  | AstNumberLiteral
  | AstStringLiteral
  | AstBinaryExpression

export enum BinaryExpressionOperator {
  // Math operators
  EXPONENT = 'EXPONENT',
  MULTIPLY = 'MULTIPLY',
  ADD = 'ADD',
  SUBTRACT = 'SUBTRACT',
  DIVIDE = 'DIVIDE',

  // Comparison operators
  EQUALS = 'EQUALS',
  NOT_EQUALS = 'NOT_EQUALS',
  LESS_THAN = 'LESS_THAN',
  LESS_THAN_EQUALS = 'LESS_THAN_EQUALS',
  GREATER_THAN = 'GREATER_THAN',
  GREATER_THAN_EQUALS = 'GREATER_THAN_EQUALS',
}
