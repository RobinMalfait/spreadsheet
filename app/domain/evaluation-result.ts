export enum EvaluationResultKind {
  ERROR = 'ERROR',
  EMPTY = 'EMPTY',
  NUMBER = 'NUMBER',
  STRING = 'STRING',
  BOOLEAN = 'BOOLEAN',
  DATETIME = 'DATETIME',
}

export type EvaluationResult =
  | { kind: EvaluationResultKind.ERROR; value: string }
  | { kind: EvaluationResultKind.EMPTY; value: string }
  | { kind: EvaluationResultKind.NUMBER; value: number }
  | { kind: EvaluationResultKind.STRING; value: string }
  | { kind: EvaluationResultKind.BOOLEAN; value: boolean }
  | { kind: EvaluationResultKind.DATETIME; value: Date; date: boolean; time: boolean }
