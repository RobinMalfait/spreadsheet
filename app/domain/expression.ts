import { type Token, TokenKind } from './tokenizer'

const UPPER_A = 65 // A
const UPPER_Z = 90 // Z

export enum AstKind {
  CELL = 'CELL',
  RANGE = 'RANGE',
  FUNCTION = 'FUNCTION',
  NUMBER_LITERAL = 'NUMBER_LITERAL',
  STRING_LITERAL = 'STRING_LITERAL',
  BINARY_EXPRESSION = 'BINARY_EXPRESSION',
}

interface Span {
  span: {
    start: number
    end: number
  }
}

export interface AstCell extends Span {
  kind: AstKind.CELL
  name: string
  loc: {
    col: number
    row: number
  }
}

export interface AstCellRange extends Span {
  kind: AstKind.RANGE
  start: AstCell
  end: AstCell
}

export interface AstFunction extends Span {
  kind: AstKind.FUNCTION
  name: string
  args: AST[]
}

export interface AstNumberLiteral extends Span {
  kind: AstKind.NUMBER_LITERAL
  value: number
}

export interface AstStringLiteral extends Span {
  kind: AstKind.STRING_LITERAL
  value: string
}

export enum BinaryExpressionOperator {
  // Math operators
  EXPONENT = 'EXPONENT',
  MULTIPLY = 'MULTIPLY',
  ADD = 'ADD',
  SUBTRACT = 'SUBTRACT',
  DIVIDE = 'DIVIDE',

  // Comparison operators
  EQUALS = 'EQUALS',
  NOT_EQUALS = 'NOT_EQUALS',
  LESS_THAN = 'LESS_THAN',
  LESS_THAN_EQUALS = 'LESS_THAN_EQUALS',
  GREATER_THAN = 'GREATER_THAN',
  GREATER_THAN_EQUALS = 'GREATER_THAN_EQUALS',
}

export interface AstBinaryExpression extends Span {
  kind: AstKind.BINARY_EXPRESSION
  operator: BinaryExpressionOperator
  lhs: AST
  rhs: AST
}

export type AST =
  | AstCell
  | AstCellRange
  | AstFunction
  | AstNumberLiteral
  | AstStringLiteral
  | AstBinaryExpression

export function parseExpression(tokens: Token[]): AST {
  return new ExpressionParser(tokens).parseExpression(0)
}

class ExpressionParser {
  constructor(private tokens: Token[]) {
    this.tokens = tokens.slice()
  }

  isBinaryOperator(token: Token) {
    return (
      // Math operators
      token.kind === TokenKind.EXPONENT ||
      token.kind === TokenKind.ASTERISK ||
      token.kind === TokenKind.PLUS ||
      token.kind === TokenKind.MINUS ||
      token.kind === TokenKind.DIVIDE ||
      // Comparison operators
      token.kind === TokenKind.LESS_THAN ||
      token.kind === TokenKind.EQUALS ||
      token.kind === TokenKind.GREATER_THAN ||
      token.kind === TokenKind.LESS_THAN_EQUALS ||
      token.kind === TokenKind.GREATER_THAN_EQUALS ||
      token.kind === TokenKind.NOT_EQUALS
    )
  }

  toBinaryOperator(token: Token) {
    switch (token.kind) {
      // Math operators
      case TokenKind.EXPONENT:
        return BinaryExpressionOperator.EXPONENT
      case TokenKind.ASTERISK:
        return BinaryExpressionOperator.MULTIPLY
      case TokenKind.PLUS:
        return BinaryExpressionOperator.ADD
      case TokenKind.MINUS:
        return BinaryExpressionOperator.SUBTRACT
      case TokenKind.DIVIDE:
        return BinaryExpressionOperator.DIVIDE

      // Comparison operators
      case TokenKind.LESS_THAN:
        return BinaryExpressionOperator.LESS_THAN
      case TokenKind.LESS_THAN_EQUALS:
        return BinaryExpressionOperator.LESS_THAN_EQUALS
      case TokenKind.EQUALS:
        return BinaryExpressionOperator.EQUALS
      case TokenKind.NOT_EQUALS:
        return BinaryExpressionOperator.NOT_EQUALS
      case TokenKind.GREATER_THAN:
        return BinaryExpressionOperator.GREATER_THAN
      case TokenKind.GREATER_THAN_EQUALS:
        return BinaryExpressionOperator.GREATER_THAN_EQUALS
      default:
        throw new Error(`Invalid binary operator: ${token.raw}`)
    }
  }

  getPrecedence(token: Token) {
    // Exponentiation
    if (token.kind === TokenKind.EXPONENT) {
      return 4
    }

    // Multiplication and division
    if (token.kind === TokenKind.ASTERISK || token.kind === TokenKind.DIVIDE) {
      return 3
    }

    // Addition and subtraction
    if (token.kind === TokenKind.PLUS || token.kind === TokenKind.MINUS) {
      return 2
    }

    // Comparison operators
    if (
      token.kind === TokenKind.EQUALS ||
      token.kind === TokenKind.NOT_EQUALS ||
      token.kind === TokenKind.LESS_THAN ||
      token.kind === TokenKind.LESS_THAN_EQUALS ||
      token.kind === TokenKind.GREATER_THAN ||
      token.kind === TokenKind.GREATER_THAN_EQUALS
    ) {
      return 1
    }

    return 0
  }

  parseIncreasingPrecedence(lhs: AST, precedence: number): AST {
    let next = this.tokens.shift()
    if (!next) return lhs

    let nextPrecedence = this.getPrecedence(next)
    if (nextPrecedence <= precedence) {
      this.tokens.unshift(next)
      return lhs
    }

    if (!this.isBinaryOperator(next)) return lhs

    let rhs = this.parseExpression(nextPrecedence)

    return {
      kind: AstKind.BINARY_EXPRESSION,
      operator: this.toBinaryOperator(next),
      lhs,
      rhs,
      span: {
        start: lhs.span.start,
        end: rhs.span.end,
      },
    }
  }

  parseExpression(precedence: number): AST {
    let lhs = this.parseLeaf()
    while (true) {
      let node = this.parseIncreasingPrecedence(lhs, precedence)
      if (node === lhs) break
      lhs = node
    }
    return lhs
  }

  parseLeaf(): AST {
    let token = this.tokens.shift()
    if (!token) throw new Error('Unexpected end of input')

    switch (token.kind) {
      case TokenKind.PLUS: {
        let next = this.tokens[0]

        if (next?.kind === TokenKind.NUMBER_LITERAL) {
          this.tokens.shift() // Consume next token
          return {
            kind: AstKind.NUMBER_LITERAL,
            value: +next.value,
            span: {
              start: token.span.start,
              end: next.span.end,
            },
          }
        }

        throw new Error(`Invalid expression, expected number literal, got ${next.raw}`)
      }

      case TokenKind.MINUS: {
        let next = this.tokens[0]

        if (next?.kind === TokenKind.NUMBER_LITERAL) {
          this.tokens.shift() // Consume next token
          return {
            kind: AstKind.NUMBER_LITERAL,
            value: -next.value,
            span: {
              start: token.span.start,
              end: next.span.end,
            },
          }
        }

        throw new Error(`Invalid expression, expected number literal, got ${next.raw}`)
      }

      case TokenKind.NUMBER_LITERAL:
        return { kind: AstKind.NUMBER_LITERAL, value: token.value, span: token.span }

      case TokenKind.STRING_LITERAL:
        return { kind: AstKind.STRING_LITERAL, value: token.value, span: token.span }

      case TokenKind.IDENTIFIER: {
        let peek = this.tokens[0]
        if (!peek) {
          // Cell reference
          return {
            kind: AstKind.CELL,
            name: token.value,
            loc: parseLocation(token.value),
            span: token.span,
          }
        }

        // Cell range
        if (peek.kind === TokenKind.COLON) {
          this.tokens.shift() // :
          let end = this.tokens.shift() // End
          if (!end || end.kind !== TokenKind.IDENTIFIER) {
            throw new Error('Invalid cell range')
          }

          return {
            kind: AstKind.RANGE,
            start: {
              kind: AstKind.CELL,
              name: token.value,
              loc: parseLocation(token.value),
              span: token.span,
            },
            end: {
              kind: AstKind.CELL,
              name: end.value,
              loc: parseLocation(end.value),
              span: end.span,
            },
            span: {
              start: token.span.start,
              end: end.span.end,
            },
          }
        }

        // Function call
        if (peek.kind === TokenKind.OPEN_PAREN) {
          // Skip open paren
          let openParen = this.tokens.shift()
          if (openParen?.kind !== TokenKind.OPEN_PAREN) {
            throw new Error('Invalid function call')
          }

          // Each argument is separated by a comma, it could be that multiple
          // tokens are required for a single expression
          let args: Token[][] = []

          // Track paren stack
          let stack = 0

          // Track tokens for next argument
          let argTokens: Token[] = []

          let closeParen: Token | null = null

          // Find closing paren
          loop: while (this.tokens.length > 0) {
            // biome-ignore lint/style/noNonNullAssertion: We already verified that there is at least 1 remaining token
            let next = this.tokens.shift()!

            switch (next.kind) {
              // Nested function call, or just a parenthesized expression
              case TokenKind.OPEN_PAREN:
                argTokens.push(next)
                stack++
                break

              // End of function call, but might not be the current function
              // call
              case TokenKind.CLOSE_PAREN:
                if (stack === 0) {
                  if (argTokens.length > 0) {
                    args.push(argTokens.splice(0))
                  }

                  closeParen = next
                  break loop
                }

                // Close paren for nested parenthesized expression
                argTokens.push(next)
                stack--
                break

              // Argument separator
              case TokenKind.COMMA:
                if (stack === 0 && argTokens.length > 0) {
                  args.push(argTokens.splice(0))
                }

                if (stack !== 0) {
                  argTokens.push(next)
                }
                break

              // Token belongs to the current argument
              default:
                argTokens.push(next)
                break
            }
          }

          if (!closeParen) {
            throw new Error('Invalid function call, missing closing paren')
          }

          return {
            kind: AstKind.FUNCTION,
            name: token.value,
            args: args.map((list) => parseExpression(list)),
            span: {
              start: token.span.start,
              end: closeParen.span.end,
            },
          }
        }

        // Cell reference
        return {
          kind: AstKind.CELL,
          name: token.value,
          loc: parseLocation(token.value),
          span: token.span,
        }
      }

      // Parenthesized expression
      case TokenKind.OPEN_PAREN: {
        let expression = this.parseExpression(0)
        let close = this.tokens.shift()

        if (close?.kind !== TokenKind.CLOSE_PAREN) {
          throw new Error(`Invalid expression, expected \`)\` got \`${token.raw}\``)
        }

        return expression
      }
    }

    throw new Error(`Invalid expression, unexpected token: ${token.raw}`)
  }
}

export type Location = {
  col: number
  row: number
}

export function parseLocation(input: string): Location {
  let idx = 0
  let char = input.charCodeAt(idx)
  do {
    char = input.charCodeAt(++idx)
  } while (char >= UPPER_A && char <= UPPER_Z)

  return {
    col: parseColNumber(input.slice(0, idx)),
    row: Number(input.slice(idx)),
  }
}

export function parseColNumber(input: string) {
  let col = 0

  for (let i = 0; i < input.length; i++) {
    let char = input.charCodeAt(i)
    if (char < UPPER_A || char > UPPER_Z) {
      throw new Error('Invalid column number')
    }
    col *= 26 // Base 26, shift left
    col += char - UPPER_A // Normalize to 0-indexed
    col += 1 // 1-indexed
  }

  return col
}

/**
 * Print AST
 */
export function printExpression(input: AST, depth = 0): string {
  switch (input.kind) {
    case AstKind.CELL:
      return input.name
    case AstKind.RANGE:
      return `${input.start.name}:${input.end.name}`
    case AstKind.FUNCTION:
      return `${input.name}(${input.args.map(printExpression).join(', ')})`
    case AstKind.NUMBER_LITERAL:
      return input.value.toString()
    case AstKind.STRING_LITERAL:
      return `"${input.value}"`
    case AstKind.BINARY_EXPRESSION:
      return depth === 0
        ? `${printExpression(input.lhs, depth + 1)} ${printBinaryOperator(input.operator)} ${printExpression(input.rhs, depth + 1)}`
        : `(${printExpression(input.lhs, depth + 1)} ${printBinaryOperator(input.operator)} ${printExpression(input.rhs, depth + 1)})`
  }
}

function printBinaryOperator(operator: BinaryExpressionOperator) {
  switch (operator) {
    // Math operators
    case BinaryExpressionOperator.EXPONENT:
      return '^'
    case BinaryExpressionOperator.ADD:
      return '+'
    case BinaryExpressionOperator.SUBTRACT:
      return '-'
    case BinaryExpressionOperator.MULTIPLY:
      return '*'
    case BinaryExpressionOperator.DIVIDE:
      return '/'

    // Comparison operators
    case BinaryExpressionOperator.EQUALS:
      return '='
    case BinaryExpressionOperator.NOT_EQUALS:
      return '<>'
    case BinaryExpressionOperator.LESS_THAN:
      return '<'
    case BinaryExpressionOperator.LESS_THAN_EQUALS:
      return '<='
    case BinaryExpressionOperator.GREATER_THAN:
      return '>'
    case BinaryExpressionOperator.GREATER_THAN_EQUALS:
      return '>='

    default:
      operator satisfies never
  }
}

export function printLocation(location: Location) {
  return `${printColNumber(location.col)}${location.row}`
}

function printColNumber(input: number) {
  let remaining = input
  let col = ''

  while (remaining > 0) {
    let mod = (remaining - 1) % 26
    col = String.fromCharCode(UPPER_A + mod) + col
    remaining = Math.floor((remaining - mod) / 26)
  }

  return col
}
