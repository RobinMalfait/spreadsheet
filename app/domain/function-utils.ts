import type { EvaluationResult } from '~/domain/evaluation-result'
import { parse } from '~/domain/signature/parser'
import { tokenize } from '~/domain/signature/tokenizer'
import { validate } from '~/domain/signature/validate'

export function expose<
  T extends EvaluationResult,
  R extends EvaluationResult | EvaluationResult[],
>(signature: string, handle: (...args: T[]) => R): (...args: T[]) => R {
  let sig = parse(tokenize(signature))

  return Object.assign(
    (...args: T[]) => {
      let errors = validate(sig, args)
      if (errors) return errors as R

      // Handle
      return handle(...args) as R
    },
    { signature: sig },
  )
}
