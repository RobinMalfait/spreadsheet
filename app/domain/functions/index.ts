export * from './binary'
export * from './date'
export * from './logic'
export * from './lookup'
export * from './math'
export * from './privileged'
export * from './sequence'
export * from './statistics'
export * from './text'
export * from './types'
