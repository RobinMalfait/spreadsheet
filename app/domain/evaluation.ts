import { format } from 'date-fns'
import { type AST, AstKind, BinaryExpressionOperator } from '~/domain/ast'
import * as functions from '~/domain/functions'
import type { Spreadsheet } from '~/domain/spreadsheet'
import { expandRange } from '~/domain/walk-ast'

export enum EvaluationResultKind {
  NUMBER = 'NUMBER',
  STRING = 'STRING',
  BOOLEAN = 'BOOLEAN',
  DATETIME = 'DATETIME',
}

export type EvaluationResult =
  | { kind: EvaluationResultKind.NUMBER; value: number }
  | { kind: EvaluationResultKind.STRING; value: string }
  | { kind: EvaluationResultKind.BOOLEAN; value: boolean }
  | { kind: EvaluationResultKind.DATETIME; value: Date; date: boolean; time: boolean }

export function evaluateExpression(
  ast: AST,
  spreadsheet: Spreadsheet,
): EvaluationResult[] {
  switch (ast.kind) {
    case AstKind.NUMBER_LITERAL:
      return [{ kind: EvaluationResultKind.NUMBER, value: ast.value }]

    case AstKind.STRING_LITERAL: {
      if (ast.value.trim() === '') {
        return [{ kind: EvaluationResultKind.STRING, value: ast.value }]
      }

      // Try to coerce the string to a number, and if it works, return a number
      // instead of a string
      let asNumber = Number(ast.value)
      if (!Number.isNaN(asNumber)) {
        return [{ kind: EvaluationResultKind.NUMBER, value: asNumber }]
      }

      return [{ kind: EvaluationResultKind.STRING, value: ast.value }]
    }

    case AstKind.CELL:
      return spreadsheet.evaluate(ast.name)

    case AstKind.RANGE: {
      let out = []
      for (let cell of expandRange(ast)) {
        for (let child of spreadsheet.evaluate(cell)) {
          out.push(child)
        }
      }
      return out
    }

    case AstKind.BINARY_EXPRESSION: {
      let lhs = evaluateExpression(ast.lhs, spreadsheet)
      let rhs = evaluateExpression(ast.rhs, spreadsheet)

      if (lhs.length !== 1 || rhs.length !== 1) {
        throw new Error('Expected a single result from each side')
      }

      let left = lhs[0]
      let right = rhs[0]

      if (
        left?.kind === EvaluationResultKind.NUMBER &&
        right?.kind === EvaluationResultKind.NUMBER
      ) {
        switch (ast.operator) {
          // Math operators
          case BinaryExpressionOperator.EXPONENT:
            return [functions.POWER(left, right)]
          case BinaryExpressionOperator.MULTIPLY:
            return [functions.MULTIPLY(left, right)]
          case BinaryExpressionOperator.DIVIDE:
            return [functions.DIVIDE(left, right)]
          case BinaryExpressionOperator.ADD:
            return [functions.SUM(left, right)]
          case BinaryExpressionOperator.SUBTRACT:
            return [functions.SUBTRACT(left, right)]

          // Comparison operators
          case BinaryExpressionOperator.EQUALS:
            return left.value === right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.NOT_EQUALS:
            return left.value !== right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.LESS_THAN:
            return left.value < right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.LESS_THAN_EQUALS:
            return left.value <= right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.GREATER_THAN:
            return left.value > right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.GREATER_THAN_EQUALS:
            return left.value >= right.value ? [functions.TRUE()] : [functions.FALSE()]
          default:
            ast.operator satisfies never
        }
      }

      throw new Error(`Invalid operation \`${ast.kind}\``)
    }

    case AstKind.FUNCTION: {
      if (!Object.hasOwn(functions, ast.name)) {
        throw new Error(`Unknown function \`${ast.name}\``)
      }

      let fn = functions[ast.name as keyof typeof functions]
      let args = ast.args.flatMap((arg) => evaluateExpression(arg, spreadsheet))
      // @ts-expect-error Some functions have a different arity, but we're not
      // checking that here.
      let result = fn(...args)
      if (Array.isArray(result)) {
        return result
      }
      return [result]
    }
  }
}

export function printEvaluationResult(result: EvaluationResult): string {
  switch (result.kind) {
    case EvaluationResultKind.NUMBER:
      return result.value.toString()
    case EvaluationResultKind.STRING:
      return result.value
    case EvaluationResultKind.BOOLEAN:
      return result.value ? 'TRUE' : 'FALSE'
    case EvaluationResultKind.DATETIME:
      if (result.date && result.time) return format(result.value, 'yyyy-MM-dd HH:mm:ss')
      if (result.date) return format(result.value, 'yyyy-MM-dd')
      return format(result.value, 'HH:mm:ss')
    default:
      result satisfies never
      return ''
  }
}
