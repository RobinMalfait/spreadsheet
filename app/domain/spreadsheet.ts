import { type AST, AstKind } from '~/domain/ast'
import { type EvaluationResult, evaluateExpression } from '~/domain/evaluation'
import { parse } from '~/domain/expression'
import * as functions from '~/domain/functions'
import { tokenize } from '~/domain/tokenizer'
import { WalkAction, walk } from '~/domain/walk-ast'
import { DefaultMap } from '~/utils/default-map'

export enum ComputationResultKind {
  VALUE = 'VALUE',
  ERROR = 'ERROR',
}

type ComputationValue = {
  kind: ComputationResultKind.VALUE
  value: EvaluationResult
}

type ComputationError = {
  kind: ComputationResultKind.ERROR
  short: string
  message: string
}

export class Spreadsheet {
  // Track each individual cell and it's contents. AST is pre-parsed.
  private cells = new Map<string, [raw: string, ast: AST]>()

  // Track all dependencies for each cell.
  private _dependencies = new DefaultMap<string, Set<string>>(() => new Set<string>())

  private evaluationCache = new Map<string, EvaluationResult[]>()

  get cellNames() {
    return Array.from(this.cells.keys())
  }

  has(cell: string): boolean {
    return this.cells.has(cell)
  }

  get(cell: string): string {
    let result = this.cells.get(cell)
    if (result) {
      return result[0]
    }
    return ''
  }

  set(cell: string, value: string) {
    // Reset state
    this.cells.delete(cell)

    // Clear existing dependencies for this cell
    let dependencies = this._dependencies.get(cell)
    dependencies.clear()

    // Clear the full evaluation cache. Very naive, but let's re-compute
    // everything the moment _something_ changes.
    this.evaluationCache.clear()

    // Value must be filled in
    if (value.trim() === '') return

    // Expression should exist
    let expression = value[0] === '=' ? value.slice(1) : `"${value}"`
    if (expression.trim() === '') return

    let tokens = tokenize(expression)
    let ast = parse(tokens)

    // Track all references in the AST
    walk([ast], (node) => {
      if (node.kind === AstKind.CELL) {
        dependencies.add(node.name)
      }

      return WalkAction.Continue
    })

    this.cells.set(cell, [value, ast])
  }

  functions() {
    let collator = new Intl.Collator('en', { sensitivity: 'base' })
    return Object.keys(functions).sort(collator.compare)
  }

  dependencies(cell: string): Set<string> {
    return new Set(this._dependencies.get(cell))
  }

  compute(cell: string): ComputationValue | ComputationError | null {
    try {
      let evaluationResult = this.evaluate(cell)
      if (evaluationResult.length > 1) {
        throw Object.assign(
          new Error(`Expected a single result, got ${evaluationResult.length}`),
          { short: '#VALUE!' },
        )
      }

      if (evaluationResult.length === 1) {
        return {
          kind: ComputationResultKind.VALUE,
          // biome-ignore lint/style/noNonNullAssertion: we already verified that the evaluationResult has at least one element.
          value: evaluationResult[0]!,
        }
      }

      return null
    } catch (err: unknown) {
      return {
        kind: ComputationResultKind.ERROR,
        // @ts-expect-error This is fineâ€¦
        short: err?.short ?? 'Error',
        message: (err as Error).message,
      }
    }
  }

  evaluate(cell: string): EvaluationResult[] {
    let result = this.cells.get(cell)
    if (!result) return []

    let cached = this.evaluationCache.get(cell)
    if (cached) return cached

    // TODO: Should this be moved to the `set` method?
    if (result[1].kind === AstKind.RANGE) {
      throw Object.assign(new Error('Cannot reference a range to a cell'), {
        short: '#VALUE!',
      })
    }

    // Verify references
    let dependencies = this._dependencies.get(cell)
    if (dependencies.size > 0) {
      let handled = new Set<string>()
      let todo = Array.from(dependencies)

      // Track where we came from
      let previous = cell

      while (todo.length > 0) {
        // biome-ignore lint/style/noNonNullAssertion: We verified that there is at least a single element in the while condition above.
        let next = todo.shift()!

        // No need to re-verify cells we've already handled
        if (handled.has(next)) continue
        handled.add(next)

        if (next === cell) {
          throw Object.assign(
            new Error(`Circular reference detected in cell ${previous}`),
            { short: '#REF!' },
          )
        }

        if (this.cells.has(next)) {
          // Track where we came from
          previous = next

          // Check transitive dependencies
          for (let other of this._dependencies.get(next)) {
            if (!handled.has(other)) {
              todo.push(other)
            }
          }
        }
      }
    }

    let out = evaluateExpression(result[1], this)
    this.evaluationCache.set(cell, out)
    return out
  }
}
