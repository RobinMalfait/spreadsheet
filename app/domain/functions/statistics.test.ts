import { describe, expect, it } from 'vitest'
import { Spreadsheet } from '~/domain/spreadsheet'
import { visualizeSpreadsheet } from '~/test/utils'

describe('AVERAGE()', () => {
  it('should calculate the average of a set of numbers', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=2')
    spreadsheet.set('B1', '=3')
    spreadsheet.set('C1', '=4')
    spreadsheet.set('A2', '=AVERAGE(A1:C1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───┬───┬───┐
      │   │ A │ B │ C │
      ├───┼───┼───┼───┤
      │ 1 │ 2 │ 3 │ 4 │
      ├───┼───┼───┼───┤
      │ 2 │ 3 │   │   │
      └───┴───┴───┴───┘
      "
    `)
  })

  it('should ignore non-number types', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=2')
    spreadsheet.set('B1', '=3')
    spreadsheet.set('C1', '=TRUE()')
    spreadsheet.set('A2', '=AVERAGE(A1:C1, 4, NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───┬───┬──────┐
      │   │ A │ B │ C    │
      ├───┼───┼───┼──────┤
      │ 1 │ 2 │ 3 │ TRUE │
      ├───┼───┼───┼──────┤
      │ 2 │ 3 │   │      │
      └───┴───┴───┴──────┘
      "
    `)
  })
})
