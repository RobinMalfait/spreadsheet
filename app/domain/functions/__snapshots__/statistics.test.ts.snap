// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@example tests > AVERAGE(1, 3, 2, 5, 4) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 3 │
└───┴───┘
"
`;

exports[`@example tests > COUNT(1, 2, TRUE(), "hello world", 3.2) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 3 │
└───┴───┘
"
`;

exports[`@example tests > MAX(1, 3, 2, 5, 4) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 5 │
└───┴───┘
"
`;

exports[`@example tests > MEDIAN(1, 3, 2, 5, 4) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 3 │
└───┴───┘
"
`;

exports[`@example tests > MIN(5, 4, 1, 2, 3) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 1 │
└───┴───┘
"
`;

exports[`@example tests > MODE(3, 2, 1, 3, 3, 4, 5, 8, 9, 1) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 3 │
└───┴───┘
"
`;
